text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+$ tikv-cli raw set "cncf" "linux foundation"
+Finished in 0.001s.
+$ tikv-cli raw set "tikv" "rust"
+Finished in 0.001s.
+$ tikv-cli raw get "cncf"
+"linux foundation"
+Finished in 0.002s.
+$ tikv-cli raw scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+Example of REPL raw mode:
+
+```bash
+$ tikv-cli raw
+> set "cncf" "linux foundation"
+Finished in 0.001s.
+> set "tikv" "rust"
+Finished in 0.001s.
+> get "cncf"
+"linux foundation"
+Finished in 0.001s.
+> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+REPL transactional mode, notice how the `>>` communicates that the user is in a
+transaction. Also notice how the transaction timestamp is outputted in the
+request complete messages.
+
+```bash
+$ tikv-cli txn
+> begin
+Finished in 0.000s. (txn ts: 1551216224838)
+>>set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> get "cncf"
+"linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+> commit
+Finished in 0.001s. (txn ts: 1551216224838)
+```
+
+One-off transactional mode is a bit different, as we need some way to maintain
+timestamp across commands. The user may choose how to do this, the reccomended
+way is via ENV variables, but they must do this. (See alternatives)
+
+```bash
+$ TS=$(tikv-cli txn begin)
+$ tikv-cli txn ${TS} set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} get "cncf"
+"linux foundation"
+Finished in 0.002s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} commit
+Finished in 0.004s. (txn ts: 1551216224838)
+```
+
+## Drawbacks
+
+We may not wish to need to maintain this tool, or we may wish for the community
+to maintain it.
+
+## Alternatives
+
+* *(Installation/Repo)* We could use the `tikv-client` repository and provide
+  either an opt-out or an opt-in binary from `src/bin/tikv-cli.rs`. If we chose
+  an opt-in functionality here users mgiht end up with an awkard installation
+  call.
+* (*Use `tikv-cli`*), we could use `tikv-cli` for this, but it creates as fuzzy
+  distinction between management tools and clients.
+* (*No REPL, or only REPL*). We may chose to not have a REPL mode, or we may chose
+  to *only* have a REPL mode.
+* (*A different timestamp handling for one-off mode*). We may find the way we
+  handle timestamps in transactional one-off mode to be unwieldly. Perhaps there
+  is a better way.

```

Comment:
Is this standard in other clients? (MySQL in particular). If not, I would think some way of handling multiple commands at once would work. I guess the use case we want to handle is a script which adds commands to a transaction based on the results of earlier commands in the transaction. Unless there is some clear need for that, then I'd imagine that use case is better supported using an API and we should explicitly not support it.

Reply:
Part of the reason it's defined this way is that we don't have a text representation (Eg `BEGIN SELECT * FROM foo; ELECT * FROM bar; COMMIT`)

The only other idea I had to support this use case is `tikv-client txn set foo bar then get bar` or something like that and use a seperator... But I'm not sure this is nicer.